### ISSUE TITLE
Egaint game blinking screen on Hercules video card

### ISSUE THREAD

**Issue by Fenix770** [86Box/86Box#3034]
**Describe the bug**

Th Egaint game is a tetris clone for color EGA video cards and the autor added support for Hercules and EGA monochrome video modes, in the last version dated of april, 1990  (at least the last I know). I found bug on both of these modes, but since I don't have a real PC with an Hercules monochrome card, I don't  know if is a game bug or an emulator issue.

The issue:

1. The game starts normally but only the EGA title with the brics falling can be seen. There is extra text on the screen that does not appears.
2. At the menu screen there is a point that marks the option selected it cannot be seen
3. However the most problematic part is during the game becuase the screen blinks.

There is also distorted text on the monochrome EGA video mode.

here is the HDD+CFG/NVR files:

HDD:
[8088 Micoms XL-7.zip](https://github.com/86Box/86Box/files/10437041/8088.Micoms.XL-7.zip)

NVR+CFG:
[8088 Micoms.zip](https://github.com/86Box/86Box/files/10437046/8088.Micoms.zip)

The floppy image of the game:
[EGA-INT.zip](https://github.com/86Box/86Box/files/10437188/EGA-INT.zip)

Small MP4 that shows the bug:
[8088 Micoms - 86Box 3.11 [build 4376] 2023-01-17 12-48-08.zip](https://github.com/86Box/86Box/files/10437053/8088.Micoms.-.86Box.3.11.build.4376.2023-01-17.12-48-08.zip)

Here is the mobygames site that shows how the first screen looks like:
https://www.mobygames.com/game/dos/egaint/screenshots

I thought the autor could be omited the text on monochrome modes, but on the S3 924 the game starts in EGA monochrome mode and and the text are present, the screen is blue and black instead white and black. To avoid this mis-detection I use "EGAINT C" at DOS prompt to force the color EGA mode.

![EGAINT monochrome mode on VGA](https://user-images.githubusercontent.com/31600563/212953925-21b51e94-8402-4ad6-b420-ed1de5a0f552.png)

**To Reproduce**

1. Start the VM, with an Hercules monochrome card
2. At DOS prompt type EGAIN
3. The gme starts...

**Desktop (please complete the following information):**
 - OS: Win 10 21H2
 - 86Box version: 3.11.0.4376
 - Build information: old dynarec Win32

**Comment by iamgreaser** [comment]
The source code to this game can be found here, and is licensed under the GPLv1: https://github.com/pmachapman/egaint
This uses some version of Turbo Pascal. Considering that this uses the `inline()` syntax and not the `asm` syntax for assembly, this most likely predates version 6.0 (possibly version 5.5?). The other obvious point of knowing which versions it *probably isn't* is of course the release date of 1989-1990, which would cover 6.0 at the latest.
The main graphics system is Turbo Pascal's "Graph" unit, where `DetectGraph` ends up being called.

For comparison, the behaviour being shown is the same as what happens in DOSBox-X - Hercules flickers, and anything EGA-compatible always misdetects as monochrome.

Now, this note in the `README` file stands out...

```text
The biggest improvement in version 0.94.13 is the addition (after
numerous user requests) of support (albeit largely untested) for the
Hercules and EGA monochrome graphics adapters.  SINCE I NEITHER OWN NOR
HAVE ACCESS TO EITHER OF THESE GRAPHICS ADAPTERS, I CAN PRACTICALLY
GUARANTEE BUGS FOR HERCULES AND EGA MONOCHROME USERS.  Please try to
report these bugs as time permits.  (If you have a color EGA or VGA,
never mind, this doesn't apply to you.)
```

So there's a good chance that the Hercules behaviour is what would happen on real hardware.
But everything being misdetected as monochrome on EGA and compatible stuff is what currently stumps me.

**Comment by iamgreaser** [comment]
Well, the mystery of the misdetected video mode has been mostly resolved - it's detecting video modes correctly, but the monitor colourness (?) is being stored in `egaint.rc`. Whenever you run it with Hercules graphics, it resets this setting to mono.

As for this:

> There is also distorted text on the monochrome EGA video mode.

This may have been fixed at some point?

So now we're at the final mystery: What causes the Hercules mode to blink? It's got something to do with page flipping, but the code does appear to be doing the right thing at least in some places...

**Comment by iamgreaser** [comment]
Alright, I managed to get it building with Turbo Pascal 5.5 and found out what's going on. I did all the testing for this in DOSBox-X considering that the behaviour is the same as 86Box.

The game uses ClearDevice when it wants to clear a page on the screen. The TP5.0 reference guide says:

> Clears the graphics screen and prepares it for output.
>
> *ClearDevice* moves the current pointer to (0,0), clears the screen using the background color set by *SetBkColor*, and prepares it for output.

Alternatively, one can use ClearViewPort:

> Clears the current viewport.
>
> Sets the fill color to the background color (*Palette*[0]), calls *Bar*, and moves the current pointer to (0,0).

And to clarify, Bar draws a filled rectangle.

What I suspect is happening is that the Hercules driver interprets a call to ClearDevice as one to clear *both* pages, while the EGA and VGA drivers seem to only clear the *active* page.

Replacing all calls to ClearDevice with ClearViewPort fixes the blinking.

With that said, I can say the following:

- EGA mono, distorted text: Appears to be fixed at this point.
- Colour EGA/VGA/SVGA going into monochrome mode: This is stored in `egaint.rc`. After launching `egaint c`, one can launch it as `egaint` without the `c` and it'll still be in colour mode. Alternatively, one can delete `egaint.rc`.
- Hercules blinking: Potentially a bug in `ClearDevice` in the Turbo Pascal 5.x Hercules driver.

This issue can be closed.
